Feat/recommender engagement tracking

scopalaffairs opened the pull request:
# People Discovery and Recommendation Components

Services and Views for

* Trending Users
* Recommended Users
* Similar Users

Engagement Tracker

for Recommender Components

charleslavon commented on src/Recommender/Engagement/FollowButtonTracked.jsx:
```diff
@@ -0,0 +1,28 @@
+return (
+  <div onClick={handleClick}>
+    <Widget
+      src="${REPL_ACCOUNT}/widget/Recommender.Service.EngagementTracker"
```
@roshaans This PR appears to introduce an alternative to the product analytics events recorded in`near-discovery`. Can you help me understand when one should use this approach to record product usage data, versus the approach in `near-discovery`, versus the `data-attr` approach that you shared last week?

scopalaffairs replied:
Hi @charleslavon,

Thank you for bringing this up. Allow me to clarify, please:

**This PR's Approach**: I was specifically tasked to develop a solution focused on click analytics for the recommender components. This method was developed independently of the `data-attr` approach, with no coordination between them.
For testing purposes, I've set up a separate RudderStack account. Additionally, there's also a dedicated source for the recommender features in production. I believe this separation can be beneficial, but I'm open to feedback on this structure.

**data-attr Approach**: I was only recently introduced to this method, last Thursday. Although it's promising, our immediate priority is to deploy engagement tracking for recommendations quickly.

Given the urgency and based on discussions with Pavel, we chose to proceed with the solution in this PR. We're definitely open to consolidating or refining our approach in the future based on feedback and overall strategy.

I hope that clarifies things. Please let me know if you have more questions.

Best,
Daniel

roshaans replied:
thank you for the response @scopalaffairs. 

We currently still do not have a way to track engagement of buttons and components on the BOS yet. Well, until this  [PR 131](https://github.com/NearSocial/VM/pull/131) is merged. 

So till then, we have to use this approach that Daniel worked on.

charleslavon replied:
thanks for the detail @scopalaffairs and @roshaans 

Will this data be piped into Pagoda's Rudderstack instance, or is the intent to allow gateway operators to specify a rudderstack destination?

scopalaffairs replied:
@charleslavon currently, it's piped into a specific rudderstack source that shares the prod rudderstack DB deltalake destination with other sources.

scopalaffairs commented on src/Recommender/Account/AccountProfileLargeCard.jsx:
```diff
@@ -0,0 +1,243 @@
+        }}
+      />
+
+    {tags.length > 0 ? (
```
yup! nice. 👍

scopalaffairs commented on src/Recommender/Account/AccountProfileLargeCard.jsx:
```diff
@@ -0,0 +1,243 @@
+
+return (
+  <LargeCard>
+    <Avatar href={profileUrl}>
```
Thank you for pointing that out, @shelegdmitriy. I've looked into the DIG.Avatar component. While it is true that the DIG.Avatar wraps an image similarly, as I am doing in the tracker component, `Recommender.Engagement.ImageTracked` in this case, I am not entirely sure about the benefits of a rafactoring.
 
As for line 140, it is indeed a styled `<a> tag`. I based its pattern and naming convention on what I observed in `src.AccountProfile`. I hope this provides clarity and satisfies this comment in your change request. 
If there are any further suggestions or concerns, please let me know.

roshaans commented on src/Recommender/Engagement/ImageTracked.jsx:
```diff
@@ -0,0 +1,31 @@
+return (
+  <div onClick={handleClick}>
+    <Widget
+      src="${REPL_ACCOUNT}/widget/Recommender.Service.EngagementTracker"
```
@scopalaffairs Did you explore using `Vm.Require` introduced by this  [PR on the VM?](https://github.com/NearSocial/VM/pull/61)

Just curious.

scopalaffairs replied:
@roshaans, I did take a look at it and find it an appealing concept, but did not test it, and since wrapping the target components in the service already worked as expected, I decided to stick with a working solution.

roshaans commented on src/Recommender/Service/EngagementTracker.jsx:
```diff
@@ -0,0 +1,37 @@
+
+const trackEngagement = () => {
+  const payload = {
+    event: eventType,
```
Here is the failed request payload. 

![image](https://github.com/near/near-discovery-components/assets/25015977/ab18da25-93e2-44a5-9e2a-5c277b6f09d5)

roshaans replied:
This is the error message it failed with: 
![image](https://github.com/near/near-discovery-components/assets/25015977/824544a0-3887-42f1-b3dd-6b70620d26e8)

For context, this is what a good request payload looks like on near.org.
![image](https://github.com/near/near-discovery-components/assets/25015977/f596d515-d58e-407e-b689-c8e83d353381)

You can perhaps retry by putting a `userId` and see if that resolves the errors.

scopalaffairs replied:
aha. interesting. Posting to the test account, I get an empty {} as error, but the event is still processed. Will update the payload accordingly 👍
Thanks

scopalaffairs replied:
@roshaans hm, I am not sure, what's wrong here, I get a 400 with `Request neither has anonymousId nor userId` even after fixing the required common fields in the payload, posting to the prod source, while I can post with a 200 OK to the test environment, still.

Test:
<img width="1115" alt="Screenshot 2023-10-04 at 12 13 49" src="https://github.com/near/near-discovery-components/assets/17599206/34357ec4-6b93-46f7-9a90-5b7f11ecd7b7">

Prod:
<img width="1107" alt="Screenshot 2023-10-04 at 12 17 18" src="https://github.com/near/near-discovery-components/assets/17599206/3b5b4e7a-dc8f-40f4-bf44-a12908d9d8ad">

Looking at the settings in rudderstack, the source seems fine and similar to the test one. Any insights?

roshaans reviewed (changes requested):
re-check logic and events being sent for analytics purpose + a comment on using `InfiniteScroll` components. 

Otherwise, great job!

scopalaffairs commented on src/PeoplePage.jsx:
```diff
@@ -275,36 +278,77 @@ return (
+        </Items>
+      )}
+
+    {!context.accountId && state.selectedTab == "trending" && (
```
@roshaans One is for logged in users, one for logged out ux. Is there a better of doing that?

scopalaffairs commented on src/PeoplePage.jsx:
```diff
@@ -275,36 +278,77 @@ return (
+      state.selectedTab == "following" ||
+      state.selectedTab == "followers") &&
+      showLoadMoreButton && (
+        <Button
```
I have experimented with the infinite-scroll feature. Although its implementation is feasible, it currently doesn't support lazy loading. Notably, some accounts' results for recommended users exceed 3MB. I believe this data must be entirely pre-loaded for the infinite scroll to function correctly.
It's concerning that despite the data being appended at the bottom, a potential race condition in state handling might be causing the view to unexpectedly scroll to the top. I am yet to determine the root cause of this behavior.

scopalaffairs commented on src/Recommender/Account/AccountProfileOverlay.jsx:
```diff
@@ -0,0 +1,293 @@
+const accountId = props.accountId;
```
Using recommender props or conditions within `src` components can clutter the code. I favor bundling functional parts, like for the tracker feature.

For AccountProfileOverlay, consider this code:
`
{props.recommender ? <Widget src="TrackFollowButton"> : <Widget src="NormalFollowButton">}
`
I believe outsourcing is simpler and reduces confusion, when all adjusted parts are in one feature directory. 

Likewise, for AccountProfileSidebar, I advocate for keeping functionality and features aligned, as seen in Search.Account.

I think this approach not only streamlines the code but also eases component maintenance. Though it's valid that we might have multiple sources, especially with latest verifications badges introduced, it's crucial to maintain clarity.

Unless we're integrating imports and service modules, finding a universally satisfying and efficient pattern remains a challenge, I think.

scopalaffairs commented on src/Recommender/Account/AccountProfileSidebar.jsx:
```diff
@@ -0,0 +1,183 @@
+const accountId = props.accountId || context.accountId;
```
See answer above

scopalaffairs commented:
@roshaans I refactored this PR quite heavily to fix the loadMore issue, by splitting concerns and removing a lot of the props drilling patterns. This improves each component's readability, on the other hand, I introduced a lot of duplicate code for fetching the data, see Recommender.Views. However appending the data now works as expected.

Also, I integrated a callback handler that removes a profile from the list onFollow in LatestPeople, best viewed via HomePage.
I found that CommitButton allows a callback function onCommit and onClick.
Thus, I wanted to change the tracking functionality to be triggered onCommit, which worked fine for PeoplePage, but I ran into issues on LatestPeople views, that's why I resorted back to using onClick.
However, this should work on both component views consistently now.

Allow me to kindly ask you again for your review after these updates.

TiffanyGYJ commented:
@roshaans 
@shelegdmitriy 
Can you please give the PR another review on if its ready to be merged? We would like this to be released by next Thursday, thank you.

roshaans commented on src/Recommender/Service/RecommendedUsers.jsx:
```diff
@@ -0,0 +1,188 @@
+      } else {
+        console.log(
+          res,
+          "Error fetching data. Try reloading the page, or no data available."
```
I see the following errors in console:

![image](https://github.com/near/near-discovery-components/assets/25015977/a4804b7b-fa4f-49a1-81f5-efee1e42ff47)

Can we remove these as I do not think they really offer any help to the user

scopalaffairs replied:
Agree, thought so as well.

roshaans commented on src/Recommender/Views/RecommendedUsersView.jsx:
```diff
@@ -0,0 +1,154 @@
+          <></>
+        ) : (
+          <CategoryHeader>
+            <H3>Similar to you</H3>
```
![image](https://github.com/near/near-discovery-components/assets/25015977/62c178dc-bfbc-45fd-8797-dcdc0cbb4732)

I believe I follow enough users so i should be seeing profiles in this section. 

When  I click into `View All` I do end up seeing profiles however

![image](https://github.com/near/near-discovery-components/assets/25015977/8ba1f174-3b3f-4e1b-a4ca-e4d9f8f5fb7e)

roshaans commented on src/Recommender/Account/AccountProfileLargeCard.jsx:
```diff
@@ -0,0 +1,245 @@
+    )}
+
+    {props.following !== null &&
+    props.followers !== null &&
```
not quite sure where the 22 is coming from. Some context on this number would be nice.

![image](https://github.com/near/near-discovery-components/assets/25015977/d9302db8-e667-4f84-bcfd-12e1fd9217b5)

scopalaffairs replied:
@roshaans 
context is: debugging statement and I ran out of recommendations, thus mixing up the condition!
fixed! 
😅 
Thanks for catching that!

roshaans reviewed (approved):
Awesome work! Just some small nitty gritty details left. Tracking also works as expected. 

Approving this as I think none of the issues I mentioned are blockers, but you can treat them as suggestions and fix them if you think they are necessary.

shelegdmitriy commented on src/PeoplePage.jsx:
```diff
@@ -275,36 +278,81 @@ return (
+        </Items>
+      )}
+
+    {!context.accountId && state.selectedTab == "trending" && (
+      <Widget
+        src="${REPL_ACCOUNT}/widget/Recommender.Views.TrendingUsersView"
+        props={{ currentPage: state.currentPage, fromContext: fromContext }}
+      />
+    )}
+
+    {context.accountId && state.selectedTab == "trending" && (
+      <Widget
+        src="${REPL_ACCOUNT}/widget/Recommender.Views.TrendingUsersView"
+        props={{ currentPage: state.currentPage, fromContext: fromContext }}
+      />
```
Can't we just render it in this way:

```suggestion
    {state.selectedTab == "trending" && (
      <Widget
        src="${REPL_ACCOUNT}/widget/Recommender.Views.TrendingUsersView"
        props={{ currentPage: state.currentPage, fromContext: fromContext }}
      />
```
?

scopalaffairs replied:
One of the views is for logged out user experience. 
I think, it won't render if not mentioned explicitly

shelegdmitriy commented on src/Recommender/Account/AccountProfileLargeCard.jsx:
```diff
@@ -0,0 +1,245 @@
+
+    {props.following !== null &&
+    props.followers !== null &&
+    props.likers !== null ? (
```
At the end of this condition you're returning an empty tags

```jsx
<Scores>
  <Score></Score>
</Scores>
```

Why we can't do this
```suggestion
    props.likers !== null && (
```
instead?

scopalaffairs replied:
True, what you say, a matter of card height.

shelegdmitriy commented on src/Recommender/Engagement/FollowButtonTracked.jsx:
```diff
@@ -0,0 +1,30 @@
+      }}
+    />
+    <Widget
+      src={"${REPL_ACCOUNT}/widget/FollowButton"}
```
```suggestion
      src="${REPL_ACCOUNT}/widget/FollowButton"
```

shelegdmitriy commented on src/Recommender/Engagement/ProfileInfoTracked.jsx:
```diff
@@ -0,0 +1,127 @@
+`;
+
+return (
+  <div
+    onClick={handleClick}
+    style={{
+      display: "flex",
+      flexDirection: "column",
+      justifyContent: "space-between",
+    }}
+  >
```
Make this `div` as styled component with it's own styles

shelegdmitriy commented on src/Recommender/Service/RecommendedUsers.jsx:
```diff
@@ -0,0 +1,195 @@
+            <div>
+              Follow More Users to Unlock More Personalized Recommendations, See
+              Who’s
+              <a href="https://near.org/${REPL_ACCOUNT}/widget/PeoplePage?tab=trending">
```
I think you can use `REPL_NEAR_URL` here:

```suggestion
              <a href="https://${REPL_NEAR_URL}/${REPL_ACCOUNT}/widget/PeoplePage?tab=trending">
```

shelegdmitriy commented on src/Recommender/Service/RecommendedUsers.jsx:
```diff
@@ -0,0 +1,195 @@
+const accountId = props.accountId;
+const profile = props.profile || Social.get(`${accountId}/profile/**`, "final");
+const tags = Object.keys(profile.tags || {});
+const profileUrl = `#/near/widget/ProfilePage?accountId=${accountId}`;
```
```suggestion
const profileUrl = `#/${REPL_ACCOUNT}/widget/ProfilePage?accountId=${accountId}`;
```

shelegdmitriy commented on src/Recommender/Service/RecommendedUsers.jsx:
```diff
@@ -0,0 +1,195 @@
+      });
+    }
+  } catch (error) {
+    console.log(error.message);
```
Let's be more specific with errors:

```suggestion
    console.error("Error on fetching recommended users: ", error.message);
```

shelegdmitriy commented on src/Recommender/Views/FriendsOfFriends.jsx:
```diff
@@ -0,0 +1,171 @@
+      });
+    }
+  } catch (error) {
+    console.log(error.message);
```
Let's be more specific with errors:

```suggestion
    console.error("Error on fetching recommended users: ", error.message);
```

shelegdmitriy commented on src/Recommender/Views/FriendsOfFriends.jsx:
```diff
@@ -0,0 +1,171 @@
+        displayedUsers.map((user, index) => (
+          <Profile key={index}>
+            <Widget
+              src="scopalaffairs.near/widget/Recommender.Account.AccountProfileLargeCard"
```
It seems @scopalaffairs you're already created such component and forgot to change src, right?

```suggestion
              src="${REPL_ACCOUNT}/widget/Recommender.Account.AccountProfileLargeCard"
```

shelegdmitriy commented on src/Recommender/Views/FriendsOfFriends.jsx:
```diff
@@ -0,0 +1,171 @@
+        </>
+      ) : (
+        displayedUsers.map((user, index) => (
+          <Profile key={index}>
```
`index` as key is not really helpful

shelegdmitriy commented on src/Recommender/Views/TrendingUsersView.jsx:
```diff
@@ -0,0 +1,199 @@
+        {state.userData.map((user, index) => (
+          <Profile key={index}>
+            <Widget
+              src="scopalaffairs.near/widget/Recommender.Account.AccountProfileLargeCard"
```
```suggestion
              src="${REPL_ACCOUNT}/widget/Recommender.Account.AccountProfileLargeCard"
```

shelegdmitriy reviewed (changes requested):
Overall looks good @scopalaffairs. Please address comments or correct me if I'm wrong. Ping me so I'll take a look at it once again.

scopalaffairs commented on src/Recommender/Account/AccountProfileLargeCard.jsx:
```diff
@@ -0,0 +1,245 @@
+        />
+      </Button>
+    ) : (
+      <Button>
+        <CurrentUserProfile></CurrentUserProfile>
+      </Button>
```
To control height in the component.

shelegdmitriy commented:
@scopalaffairs just for future reference, take a look at our DIG component library: https://near.org/near/widget/DIG.OverviewPage

In general, let's try and use these UI components going forward to help achieve a more consistent look and feel throughout the app. If you don't have the time right now, feel free to merge this PR and refactor to using the DIG components at a later time 🙂 If you have any questions or run into issues with a DIG component, please reach out to me or @calebjacob

scopalaffairs commented on src/Recommender/Service/EngagementTracker.jsx:
```diff
@@ -0,0 +1,50 @@
+// const dataplane = "https://neardanieossax.dataplane.rudderstack.com"; //test
+const dataplane = "https://near.dataplane.rudderstack.com"; //prod
+const uri = "/v1/track";
+const api_url = `${dataplane}${uri}`;
+// const auth = "Basic MlVvMlBYSE9UdzJjUWRucThJUWJQTG9DOG5mOg=="; //test
```
we can remove it, certainly, @shelegdmitriy

scopalaffairs commented on src/Recommender/Service/RecommendedUsers.jsx:
```diff
@@ -0,0 +1,195 @@
+return (
+  <RecommendedUsers>
+    {state.isLoading && <p>Loading...</p>}
+    {state.error && <p>Error: {state.error}</p>}
```
We can move all state.error to console, sure @shelegdmitriy.
Might be helpful to give some feedback, if results are not loading?

shelegdmitriy replied:
@scopalaffairs please do the same for similar components not only for this one. I think this is the last change 🙂

scopalaffairs replied:
@shelegdmitriy, yup done.

shelegdmitriy commented:
@charleslavon I think this PR is in a pretty good state. As soon as @scopalaffairs address the last comment we can merge it, WDYT?

shelegdmitriy reviewed (approved):
LGTM!
